backend/src/main/java/ai/verta/modeldb/project/FutureProjectDAO.java

```diff
@@ -343,4 +379,358 @@ public FutureProjectDAO(
+                                    futureProjectIdsContext),
+                                executor)
+                            .thenApply(QueryFilterContext::combine, executor)
+                            .thenCompose(
+                                queryContext ->
+                                    jdbi.withHandle(
+                                        handle -> {
+                                          var sql = "select count(p.id) from project p ";
+
+                                          if (!queryContext.getConditions().isEmpty()) {
+                                            sql +=
+                                                " WHERE "
+                                                    + String.join(
+                                                        " AND ", queryContext.getConditions());
+                                          }
+
+                                          var query = handle.createQuery(sql);
+                                          queryContext.getBinds().forEach(b -> b.accept(query));
+
+                                          return query.mapTo(Long.class).one();
+                                        }),
+                                executor);
+
+                    return futureProjects
+                        .thenApply(this::sortProjectFields, executor)
+                        .thenCombine(
+                            futureCount,
+                            (projects, count) ->
+                                FindProjects.Response.newBuilder()
+                                    .addAllProjects(projects)
+                                    .setTotalRecords(count)
+                                    .build(),
+                            executor);
+                  },
+                  executor);
+            },
+            executor);
+  }
+
+  private InternalFuture<Workspace> getWorkspaceById(long workspaceId) {
+    return FutureGrpc.ClientRequest(
+        uac.getWorkspaceService()
+            .getWorkspaceById(GetWorkspaceById.newBuilder().setId(workspaceId).build()),
+        executor);
+  }
+
+  private List<Project> sortProjectFields(List<Project> projects) {
+    List<Project> sortedProjects = new LinkedList<>();
+    for (Project project : projects) {
+      var projectBuilder = Project.newBuilder(project);
+      projectBuilder
+          .clearTags()
+          .addAllTags(project.getTagsList().stream().sorted().collect(Collectors.toList()))
+          .clearAttributes()
+          .addAllAttributes(
+              project.getAttributesList().stream()
+                  .sorted(Comparator.comparing(KeyValue::getKey))
+                  .collect(Collectors.toList()))
+          .clearArtifacts()
+          .addAllArtifacts(
+              project.getArtifactsList().stream()
+                  .sorted(Comparator.comparing(Artifact::getKey))
+                  .collect(Collectors.toList()));
```

Comment:
Could we just place this responsibility into `TagsHandler` and associated? Each one returns a sorted list. That way, we ensure they will always return sorted no matter which object is being queried.

Reply:
Move this logic to individual handlers in PR: https://github.com/VertaAI/modeldb/pull/2792
Handler logic was added in PR https://github.com/VertaAI/modeldb/pull/2792 so not modified here.